### DISCUSSION TITLE
CD-like Pipeline

### DISCUSSION THREAD

**Discussion by dmalec** [jrincayc/ucblogo-code#75] [Ideas]
I've been mulling over the idea of a CD-like pipeline - the general idea would be to have a GitHub Action workflow which can build candidate release artifacts.

I have a proof of concept of what it might look like:
* Branch: https://github.com/dmalec/ucblogo-code/tree/FEAT-CD
* Pipeline: https://github.com/dmalec/ucblogo-code/actions/runs/416739610

_(I think GitHub requires you to be logged in in order to download the artifacts - otherwise they will result in 404s)_

High level what this demonstrates:
1. Build wxWidgets on Linux, OSX, & Windows
  a. This is cached after a success, that is why there are some skipped steps in the detailed view / why it's so quick.
  b. The scratch builds of each take about 5, 10, and 15 minutes respectively.
  c. The cache check tends to take 20 - 60 seconds.
  d. Cache expiration can be adjusted.
2. Cache the Inno Installer ... installer ... download for Windows (not really a speed boost, more to be a good citizen).
3. Build ucblogo on Linux - produce a `.tar.gz` and a `.pdf`
  b. The OSX and Windows builds depend on this, so they don't have to handle building the PDF.
4. Build ucblogo on OSX - produce a `.dmg`
5. Build ucblogo on Windows - produce an installer `.exe`

It's still rough in a few places:
* The Windows build statically links the 2 `.dll`s into the exe instead of including them in the `.exe`'s folder.
* I've used a bunch of `cp` commands and tweaked the inno file paths accordingly.
* There currently isn't a `.zip` artifact.
* Probably several others I'm a bit too close to the code to notice 😄 

Before going any further, I wanted to see if this sounded overall like an interesting / useful idea.

**Comment by jrincayc** [comment]
Yes, it is an interesting idea.

**Reply by jrincayc** [reply]
And if there is way to make it so anyone can download the .tar.gz, it would be very useful (but that is not required).

**Reply by dmalec** [reply]
Gotcha - I'll keep working on this one then :)

It looks like Github made a change earlier this year to prevent anonymous access to action artifacts. Sounds like that's causing 
some consternation for teams who previously relied on it to distribute nightly builds without having to mark them as full releases. I'll see if I can find an approach that allows download access to anyone without requiring the artifacts to first be included into a release.

One question - it sounds like the .tar.gz may have more significance than I had realized; when you have a chance, would you be willing to provide additional color around that point?

**Reply by jrincayc** [reply]
Generating that simplifies building on the mac, since then you just need xcode command line tools and wxWidgets.  Without that you need autoconf installed as well. (And since Gatekeeper on the mac makes installing binaries hard ...)

**Reply by dmalec** [reply]
Ah, gotcha! Thanks for the additional context, that makes sense :)

**Reply by dmalec** [reply]
Having looked around a bit, I'm still stumped on providing direct access to workflow artifacts for anonymous users. That said, it looks like one approach project are taking is to have the workflow end in publishing a release with some form of "dev latest" tag and which is marked as a _pre-release_.

From a purely technical perspective, this is doable (and provides anonymous access to the artifacts):
https://github.com/dmalec/ucblogo-code/releases/tag/dev-latest
https://github.com/dmalec/ucblogo-code/blob/FEAT-CD/.github/workflows/cd.yaml#L300

From a human perspective, I'm tentatively optimistic; but, can also see where it might cause confusion regarding releases vs development builds.

**Reply by dmalec** [reply]
This is a little embarrassing to admit; but, I just noticed the wiki pages. @jrincayc - thank you for the author credit :) I will try to align the steps in the action workflow with the steps outlined here: https://github.com/jrincayc/ucblogo-code/wiki/package_creation

**Reply by dmalec** [reply]
@jrincayc - if it will help with the release, I can rebase my CD feature branch to the release_6_2_work branch and open a PR later today - there's still a couple rough spots; but, it feels like there's a "there" there at this point.

That said, if this is a distraction from the release at this point in time, I can continue working a little bit to the side.
